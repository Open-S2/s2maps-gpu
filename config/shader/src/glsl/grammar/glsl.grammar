@top Program { unit }
@skip { space | comment }
@external tokens untilEOLToken from "./tokens" { UntilEOL }
@external tokens untilCommentCloseToken from "./tokens" { untilCommentClose }
@precedence {
  else @right,
  if @right,
  parens @left,
  subscript @left,
  unary @right,
  mul @left,
  add @left,
  shift @left,
  compare @left,
  equal @left,
  and @left,
  xor @left,
  or @left,
  andand @left,
  xorxor @left,
  oror @left,
  conditional @right,
  assign @right,
  statement @cut,
  last
}

k<term> { @specialize[@name={term}]<Identifier, term> }
d<term> { @specialize[@name={term}]<Identifier, term> }
t<term> { @specialize[@name={term}]<Identifier, term> }
r<term> { @specialize[@name={term}]<Identifier, term> }

unit { externalDeclaration }
externalDeclaration { FunctionDefinition | GlobalDeclaration | Preprocessor }

comment { commentLine | commentOpen untilCommentClose }
Preprocessor {
  hash (
    d<'pragma'> (
      d<'import'> (ImportDeclarationList 'from')? String |
      d<'export'> |
      d<'optional'> |
      d<'global'> |
      d<'optimize'> |
      d<'debug'>
      d<'STDGL'>
    ) |
    Directive
  )
  UntilEOL
}
ImportDeclarationList { '{' (ImportDeclaration comma)* (ImportDeclaration comma?) '}' }
ImportDeclaration { Identifier ('as' Identifier)? }

FunctionDefinition { FunctionPrototype compoundStatement }
FunctionPrototype { QualifiedType Identifier '(' (ParameterDeclaration (comma ParameterDeclaration)*)? ')' }
ParameterDeclaration { TypeQualifier? TypeSpecifier (Identifier ArraySpecifier?)? }

GlobalDeclaration {
  (
    FunctionPrototype |
    VariableDeclaration |
    k<"precision"> PrecisionQualifier TypeSpecifier |
    QualifiedStructDeclaration |
    QualifiedDeclaration
  ) semicolon
}

VariableDeclaration { QualifiedType (SingleDeclaration (comma SingleDeclaration)*)? }
SingleDeclaration { DeclarationSpecifier (!assign assign initializer)? }
DeclarationSpecifier { Identifier ~qualified ArraySpecifier? }

QualifiedStructDeclaration {
  TypeQualifier ~qualified (
    Identifier ~qualified (
      '{' StructDeclarationList '}' DeclarationSpecifier?
    )?
  )?
}

QualifiedDeclaration {
  TypeQualifier ~qualified (
    Identifier ~qualified (comma Identifier)*
  )?
}

LocalDeclaration {
  (
    VariableDeclaration |
    k<"precision"> PrecisionQualifier TypeSpecifier
  ) semicolon
}

PrivateIdentifier { Identifier }

TypeSpecifier { (basicType | Struct | Identifier ~qualified) ArraySpecifier? }
ArraySpecifier { (LeftBracket constantExpression RightBracket)* EmptyBrackets | (LeftBracket constantExpression RightBracket)+ EmptyBrackets? }
TypeNameList { TypeName (comma TypeName)* }
TypeName { Identifier }
QualifiedType { TypeQualifier? TypeSpecifier }

Struct { k<'struct'> Identifier? '{' StructDeclarationList '}' }
StructDeclarationList { StructDeclaration* }
StructDeclaration { TypeQualifier? TypeSpecifier structDeclaratorList semicolon }
structDeclaratorList { StructDeclarationSpecifier (comma StructDeclarationSpecifier)* }
StructDeclarationSpecifier { PrivateIdentifier ~qualified ArraySpecifier? }

initializer {
  assignmentExpression |
  '{' initializer (comma initializer)* comma? '}'
}

StatementList { (Statement | Preprocessor)* }
Statement {
  compoundStatement |
  simpleStatement
}
compoundStatement { '{' StatementList '}' }
simpleStatement {
  DeclarationStatement |
  ExpressionStatement |
  SelectionStatement |
  SwitchStatement |
  CaseLabel |
  IterationStatement |
  JumpStatement
}

DeclarationStatement { LocalDeclaration }
ExpressionStatement { expression? semicolon }
SelectionStatement { k<'if'> '(' expression ')' Statement (!else k<"else"> Statement)? }
Condition { expression | QualifiedType Identifier !assign assign initializer }
SwitchStatement { k<"switch"> '(' expression ')' '{' StatementList '}' }
CaseLabel { (k<"default"> | k<"case"> expression) ':' }
IterationStatement {
  k<"while"> '(' Condition ')' Statement |
  k<"do"> Statement k<"while"> '(' expression ')' semicolon |
  k<"for"> '(' (ExpressionStatement | DeclarationStatement) (Condition semicolon expression?) ')'
}
JumpStatement {
  k<"continue"> semicolon |
  k<"break"> semicolon |
  k<"return"> expression? semicolon |
  k<"discard"> semicolon
}
fieldSelection { PrivateIdentifier }

expression { assignmentExpression (comma assignmentExpression)* }

primaryExpression {
  Reserved? (Identifier ~qualified | Number | Parens ) 
}
Parens { '(' expression ')' }
assignmentExpression { conditionalExpression | Assignment }
constantExpression { conditionalExpression }
integerExpression { expression }
conditionalExpression {
  unaryExpression | binaryExpression | ternaryExpression
}
unaryExpression { postfixExpression | Unary }
ternaryExpression { Conditional }
binaryExpression { Binary }
postfixExpression {
  primaryExpression |
  Index |
  FunctionCall |
  Field |
  Postfix
}

Index { postfixExpression LeftBracket integerExpression RightBracket }
Field { postfixExpression '.' fieldSelection }
Postfix { postfixExpression (Inc | Dec) }
Unary { (Inc | Dec | Add | Sub | Bang | Tilde) unaryExpression }
Binary {
  conditionalExpression !mul Mul conditionalExpression |
  conditionalExpression !mul Div conditionalExpression |
  conditionalExpression !mul Mod conditionalExpression |
  conditionalExpression !add Add conditionalExpression |
  conditionalExpression !add Sub conditionalExpression |
  conditionalExpression !shift Left conditionalExpression |
  conditionalExpression !shift Right conditionalExpression |
  conditionalExpression !compare Lt conditionalExpression |
  conditionalExpression !compare Lte conditionalExpression |
  conditionalExpression !compare Gt conditionalExpression |
  conditionalExpression !compare Gte conditionalExpression |
  conditionalExpression !equal Eq conditionalExpression |
  conditionalExpression !equal Neq conditionalExpression |
  conditionalExpression !and And conditionalExpression |
  conditionalExpression !or Or conditionalExpression |
  conditionalExpression !xor Xor conditionalExpression |
  conditionalExpression !andand AndAnd conditionalExpression |
  conditionalExpression !oror OrOr conditionalExpression |
  conditionalExpression !xorxor XorXor conditionalExpression
}
Conditional { conditionalExpression !conditional '?' expression ':' assignmentExpression }
Assignment {
  unaryExpression (
    !assign (
      MulAssign |
       DivAssign |
       ModAssign |
       AddAssign |
       SubAssign |
       LeftAssign |
       RightAssign |
       AndAssign |
       XorAssign |
       OrAssign |
       assign
    )
  ) conditionalExpression
}

FunctionCall {
  (Identifier | basicType) '(' (k<'void'>? | assignmentExpression (comma assignmentExpression)*) ')'
}

TypeQualifier {
  (
    StorageQualifier |
    LayoutQualifier |
    PrecisionQualifier |
    InterpolationQualifier |
    InvariantQualifier |
    PreciseQualifier
  )+
}
StorageQualifier {
  k<'const'> |
  k<'inout'> |
  k<'in'> |
  k<'out'> |
  k<'centroid'> |
  k<'patch'> |
  k<'sample'> |
  k<'uniform'> |
  k<'buffer'> |
  k<'shared'> |
  k<'coherent'> |
  k<'volatile'> |
  k<'restrict'> |
  k<'readonly'> |
  k<'writeonly'> |
  k<'subroutine'> |
  k<'subroutine'> '(' TypeNameList ')'
}
InvariantQualifier { k<'invariant'> }
InterpolationQualifier { ( k<'smooth'> | k<'flat'> | k<'noperspective'> ) }
LayoutQualifier { ( k<'layout'> '(' LayoutQualifierId (comma LayoutQualifierId)* ')' ) }
LayoutQualifierId { PrivateIdentifier (!assign assign constantExpression)? | k<'shared'> }
PreciseQualifier { k<'precise'> }
PrecisionQualifier { ( k<'highp'> | k<'medp'> | k<'lowp'> ) }

@tokens {
  space { std.whitespace+ }

  commentLine { "//" ![\n]* }
  commentOpen { "/*" }
  
  Identifier { $[A-Za-z_] $[A-Za-z0-9_]* }
  String { '"' ( !["]+ | '\\"' )* '"' | "'" ( ![']+ | "\\'" )* "'" }

  Number {
    (
      (std.digit+ ("." std.digit*)? | "." std.digit+)
      (("e" | "E") "-"? std.digit+)?
      ('f' | 'F' | 'lf' | 'LF')?
    ) |
    std.digit+ $[uU]? |
    "0x" (std.digit | $[a-fA-F])+ $[uU]?
  }

  hash { "#" }

  Add { "+" }
  Sub { "-" }
  Mul { "*" }
  Div { "/" }
  Mod { "%" }
  Left { "<<" }
  Right { ">>" }
  And { "&" }
  Xor { "^" }
  Or { "|" }

  assign { '=' }
  AddAssign { "+=" }
  SubAssign { "-=" }
  MulAssign { "*=" }
  DivAssign { "/=" }
  ModAssign { "%=" }
  LeftAssign { "<<=" }
  RightAssign { ">>=" }
  AndAssign { "&=" }
  XorAssign { "^=" }
  OrAssign { "|=" }
  
  Inc { "++" }
  Dec { "--" }
  Bang { "!" }
  Tilde { '~' }
  
  Eq { '==' }
  Neq { '!=' }
  Lt { '<' }
  Lte { '<=' }
  Gt { '>' }
  Gte { '>=' }

  AndAnd { "&&" }
  XorXor { "^^" }
  OrOr { "||" }
  
  comma { ',' }
  semicolon { ";" }

  LeftBracket { "[" }
  RightBracket { "]" }
  EmptyBrackets { "[]"}

  @precedence {
    commentLine, Div, AndAnd, And, OrOr, Or, XorXor, Xor,
    Identifier,
    EmptyBrackets, LeftBracket
  }
}

Directive {
  d<"define"> |
  d<"undef"> |
  d<"if"> |
  d<"ifdef"> |
  d<"ifndef"> |
  d<"else"> |
  d<"elif"> |
  d<"endif"> |
  d<"error"> |
  d<"extension"> |
  d<"version"> |
  d<"line">
}

basicType {
  t<"void"> |
  t<"float"> |
  t<"double"> |
  t<"int"> |
  t<"uint"> |
  t<"atomic_uint"> |
  t<"bool"> |
  t<"mat2"> |
  t<"mat3"> |
  t<"mat4"> |
  t<"dmat2"> |
  t<"dmat3"> |
  t<"dmat4"> |
  t<"mat2x2"> |
  t<"mat2x3"> |
  t<"mat2x4"> |
  t<"dmat2x2"> |
  t<"dmat2x3"> |
  t<"dmat2x4"> |
  t<"mat3x2"> |
  t<"mat3x3"> |
  t<"mat3x4"> |
  t<"dmat3x2"> |
  t<"dmat3x3"> |
  t<"dmat3x4"> |
  t<"mat4x2"> |
  t<"mat4x3"> |
  t<"mat4x4"> |
  t<"dmat4x2"> |
  t<"dmat4x3"> |
  t<"dmat4x4"> |
  t<"vec2"> |
  t<"vec3"> |
  t<"vec4"> |
  t<"ivec2"> |
  t<"ivec3"> |
  t<"ivec4"> |
  t<"bvec2"> |
  t<"bvec3"> |
  t<"bvec4"> |
  t<"dvec2"> |
  t<"dvec3"> |
  t<"dvec4"> |
  t<"uvec2"> |
  t<"uvec3"> |
  t<"uvec4"> |
  t<"sampler3D"> |
  t<"samplerCube"> |
  t<"sampler1DShadow"> |
  t<"sampler2DShadow"> |
  t<"samplerCubeShadow"> |
  t<"sampler1DArray"> |
  t<"sampler2DArray"> |
  t<"sampler1DArrayShadow"> |
  t<"sampler2DArrayShadow"> |
  t<"samplerCubeArray"> |
  t<"samplerCubeArrayShadow"> |
  t<"isampler1D"> |
  t<"isampler2D"> |
  t<"isampler3D"> |
  t<"isamplerCube"> |
  t<"isampler1DArray"> |
  t<"isampler2DArray"> |
  t<"isamplerCubeArray"> |
  t<"usampler1D"> |
  t<"usampler2D"> |
  t<"usampler3D"> |
  t<"usamplerCube"> |
  t<"usampler1DArray"> |
  t<"usampler2DArray"> |
  t<"usamplerCubeArray"> |
  t<"sampler2DRect"> |
  t<"sampler2DRectShadow"> |
  t<"isampler2DRect"> |
  t<"usampler2DRect"> |
  t<"samplerBuffer"> |
  t<"isamplerBuffer"> |
  t<"usamplerBuffer"> |
  t<"sampler2DMS"> |
  t<"isampler2DMS"> |
  t<"usampler2DMS"> |
  t<"sampler2DMSArray"> |
  t<"isampler2DMSArray"> |
  t<"usampler2DMSArray"> |
  t<"image1D"> |
  t<"iimage1D"> |
  t<"uimage1D"> |
  t<"image2D"> |
  t<"iimage2D"> |
  t<"uimage2D"> |
  t<"image3D"> |
  t<"iimage3D"> |
  t<"uimage3D"> |
  t<"image2DRect"> |
  t<"iimage2DRect"> |
  t<"uimage2DRect"> |
  t<"imageCube"> |
  t<"iimageCube"> |
  t<"uimageCube"> |
  t<"imageBuffer"> |
  t<"iimageBuffer"> |
  t<"uimageBuffer"> |
  t<"image1DArray"> |
  t<"iimage1DArray"> |
  t<"uimage1DArray"> |
  t<"image2DArray"> |
  t<"iimage2DArray"> |
  t<"uimage2DArray"> |
  t<"imageCubeArray"> |
  t<"iimageCubeArray"> |
  t<"uimageCubeArray"> |
  t<"image2DMS"> |
  t<"iimage2DMS"> |
  t<"uimage2DMS"> |
  t<"image2DMSArray"> |
  t<"iimage2DMSArray"> |
  t<"uimage2DMSArray">
}

//Keyword {
//  k<"attribute"> |
//  k<"const"> |
//  k<"uniform"> |
//  k<"varying"> |
//  k<"buffer"> |
//  k<"shared"> |
//  k<"coherent"> |
//  k<"volatile"> |
//  k<"restrict"> |
//  k<"readonly"> |
//  k<"writeonly"> |
//  k<"layout"> |
//  k<"centroid"> |
//  k<"flat"> |
//  k<"smooth"> |
//  k<"noperspective"> |
//  k<"patch"> |
//  k<"sample"> |
//  k<"break"> |
//  k<"continue"> |
//  k<"do"> |
//  k<"for"> |
//  k<"while"> |
//  k<"switch"> |
//  k<"case"> |
//  k<"default"> |
//  k<"if"> |
//  k<"else"> |
//  k<"subroutine"> |
//  k<"in"> |
//  k<"out"> |
//  k<"inout"> |
//  k<"true"> |
//  k<"false"> |
//  k<"invariant"> |
//  k<"precise"> |
//  k<"discard"> |
//  k<"return">
//}

Reserved {
  r<"common"> |
  r<"partition"> |
  r<"active"> |
  r<"asm"> |
  r<"class"> |
  r<"union"> |
  r<"enum"> |
  r<"typedef"> |
  r<"template"> |
  r<"this"> |
  r<"resource"> |
  r<"goto"> |
  r<"inline"> |
  r<"noinline"> |
  r<"public"> |
  r<"static"> |
  r<"extern"> |
  r<"external"> |
  r<"interface"> |
  r<"long"> |
  r<"short"> |
  r<"half"> |
  r<"fixed"> |
  r<"unsigned"> |
  r<"superp"> |
  r<"input"> |
  r<"output"> |
  r<"hvec2"> |
  r<"hvec3"> |
  r<"hvec4"> |
  r<"fvec2"> |
  r<"fvec3"> |
  r<"fvec4"> |
  r<"sampler3DRect"> |
  r<"filter"> |
  r<"sizeof"> |
  r<"cast"> |
  r<"namespace"> |
  r<"using">
}
