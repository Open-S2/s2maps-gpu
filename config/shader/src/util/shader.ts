import { Tree } from '@lezer/common';
import { ASTParser, VirtualTable, SymbolTableT, ParsedBundle, ParsedModule, ParsedModuleCache, CompressedNode } from '../types';
import { formatMurmur53, toMurmur53 } from './hash';
import { PREFIX_VIRTUAL } from '../constants';

const EMPTY_LIST = [] as any[];
const EMPTY_TABLE = {} as any;

// Parse a code module into its in-memory representation
// (AST + symbol table)
export const makeLoadModule = <T extends SymbolTableT = any>(
  parseShader: (code: string) => Tree,
  makeASTParser: (code: string, tree: Tree, name?: string) => ASTParser<T>,
  compressAST: (code: string, tree: Tree, symbols?: string[]) => CompressedNode[],
  decompressAST: (nodes: CompressedNode[], symbols?: string[]) => Tree,
) => (
  code: string,
  name: string = 'main',
  entry?: string,
  compressed: boolean = false,
): ParsedModule => {
  if (code == null) throw new Error(`Shader code ${name} undefined`);
  if (typeof code !== 'string') throw new Error(`Shader code ${name} is not a string`);
  let tree = parseShader(code);

  const astParser = makeASTParser(code, tree, name);
  const table = astParser.getSymbolTable();
  const shake = astParser.getShakeTable(table);

  if (compressed) {
    const {symbols} = table;
    tree = decompressAST(compressAST(code, tree, symbols), symbols);
  }
  const hash = toMurmur53(code);

  return bindEntryPoint({name, code, hash, table, shake, tree}, entry);
}

// Use cache to load modules
export const makeLoadModuleWithCache = (
  loadModule: (code: string, name?: string, entry?: string, compressed?: boolean) => ParsedModule,
  defaultCache: ParsedModuleCache,
) => (
  code: string,
  name?: string,
  entry?: string,
  cache: ParsedModuleCache | null = defaultCache,
): ParsedModule => {
  if (!cache) return loadModule(code, name, entry, true);

  const hash = toMurmur53(code);
  const cached = cache.get(hash);
  if (cached) {
    return bindEntryPoint(cached, entry);
  }
  
  const module = loadModule(code, name, undefined, true);
  cache.set(hash, module);
  return bindEntryPoint(module, entry);
}

// Load a static (inert) module
export const loadStaticModule = (code: string, name: string, entry?: string) => {
  const hash = toMurmur53([code, entry]);
  return ({ name, code, hash, table: EMPTY_TABLE });
}

// Load a virtual (generated) module
export const loadVirtualModule = <T extends SymbolTableT = any>(
  virtual: VirtualTable,
  initTable: Partial<T> = EMPTY_TABLE,
  entry?: string,
  hash?: number,
  code?: string,
  key?: number,
) => {
  let symbols = initTable.symbols ?? EMPTY_LIST;

  code = code ?? `@virtual [${symbols.join(' ')}]`;
  hash = hash ?? toMurmur53(code);
  key  = key  ?? hash;

  const name = `${PREFIX_VIRTUAL}${formatMurmur53(key).slice(0, 6)}`;

  const table = {
    symbols,
    visibles: symbols,
    ...initTable,
  };
  return { name, code, hash, table, entry, virtual, key };
}

// Set entry point of a module, returns new bundle/module.
// Is the same instance as the original (key = old key/hash), so it merges with copies of itself.
// But is structurally different (hash = new key), so differences in links are reflected in the shader hash.
export const bindEntryPoint = <T extends ParsedBundle | ParsedModule>(bundle: T, entry?: string): T => {
  let {key, hash, module, table} = bundle as any;

  table = table ?? module?.table;
  hash = hash ?? module?.hash;

  if (entry == null && table.symbols?.includes('main')) entry = 'main';
  if (entry == null) return bundle;

  const structural = toMurmur53([hash, entry]);
  return {...bundle, entry, hash: structural, key: key ?? hash};
};
