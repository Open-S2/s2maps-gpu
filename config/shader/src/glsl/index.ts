import {
  loadModule,
  loadModuleWithCache,
  loadStaticModule,
  transpileGLSL,
  bindEntryPoint,
  defineConstants,
  makeModuleCache,

  bundleToAttribute,
  bundleToAttributes,

  glsl, float, int, uint,
} from './shader';

import {
  bindBundle,
  bindModule,
  bindingsToLinks,
  bindingToModule,
  sourceToModule,
  resolveBindings,
} from './bind';

import {
  castTo,
  swizzleTo,
} from './cast';

import {
  chainTo,
} from './chain';

import {
  diffBy,
} from './diff';

import {
  linkBundle,
  linkModule,
  linkCode,
  getPreamble,
  setPreamble,
} from './link';

import {
  makeASTParser,
  compressAST,
  decompressAST,
  rewriteUsingAST,
} from './ast';

import {
  getBundleEntry,
  getBundleHash,
  getBundleKey,
} from '../util/bundle';

import { parser } from './grammar/glsl';

export {
  loadModule,
  loadModuleWithCache,
  loadStaticModule,
  transpileGLSL,
  bindEntryPoint,
  defineConstants,
  makeModuleCache,

  bundleToAttribute,
  bundleToAttributes,

  glsl, float, int, uint,
} from './shader';

export {
  bindBundle,
  bindModule,
  bindingsToLinks,
  bindingToModule,
  sourceToModule,
  resolveBindings,
} from './bind';

export {
  castTo,
  swizzleTo,
} from './cast';

export {
  chainTo,
} from './chain';

export {
  diffBy,
} from './diff';

export {
  linkBundle,
  linkModule,
  linkCode,
  getPreamble,
  setPreamble,
} from './link';

export {
  makeASTParser,
  compressAST,
  decompressAST,
  rewriteUsingAST,
} from './ast';

export {
  getBundleEntry,
  getBundleHash,
  getBundleKey,
} from '../util/bundle';

export const GLSLLinker = {
  loadModule,
  loadModuleWithCache,
  loadStaticModule,
  transpileGLSL,
  bindEntryPoint,
  bundleToAttribute,
  bundleToAttributes,
  glsl, float, int, uint,

  defineConstants,

  linkBundle,
  linkModule,
  linkCode,
  getPreamble,
  setPreamble,

  bindBundle,
  bindModule,
  bindingsToLinks,
  bindingToModule,
  sourceToModule,
  resolveBindings,

  castTo,
  chainTo,
  diffBy,
  swizzleTo,

  makeASTParser,
  compressAST,
  decompressAST,
  rewriteUsingAST,

  makeModuleCache,

  getBundleEntry,
  getBundleHash,
  getBundleKey,

  parser,
};

export { parser } from './grammar/glsl';


export default GLSLLinker;