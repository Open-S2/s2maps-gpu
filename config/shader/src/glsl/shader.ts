import { Tree } from '@lezer/common';
import { ParsedModule, ParsedModuleCache, ShaderDefine } from './types';

import { makeLoadModule, makeLoadModuleWithCache } from '../util/shader';
import { makeBundleToAttribute, makeBundleToAttributes } from '../util/bundle';
import { makeTranspile } from '../util/transpile';

import { makeASTParser, compressAST, decompressAST } from './ast';
import { toTypeString, toTypeArgs } from './type';
import { parser } from './grammar/glsl';
import LRU from 'lru-cache';
import zip from 'lodash/zip';

export { loadStaticModule, loadVirtualModule, bindEntryPoint } from '../util/shader';

/** LRU cache for parsed shader code */
export const makeModuleCache = (options: Record<string, any> = {}) => new LRU<number, ParsedModule>({
  max: 100,
  ...options,
});

export const DEFAULT_CACHE = makeModuleCache();

/** Parse GLSL code into lezer tree */
export const parseShader = (code: string): Tree => parser.parse(code);

/** Parse a code module into its in-memory representation (AST + symbol table) */
export const loadModule = makeLoadModule(parseShader, makeASTParser, compressAST, decompressAST);

/** Use cache to load modules */
export const loadModuleWithCache = makeLoadModuleWithCache(loadModule, DEFAULT_CACHE);

/** Make GLSL definitions */
export const defineConstants = (defs: Record<string, ShaderDefine>): string => {
  const out = [];
  for (let k in defs) if (defs[k] !== false && defs[k] !== null) out.push(`#define ${k} ${defs[k]}`);
  return out.join("\n");
}

/** Make GLSL enable definitions */
export const defineEnables = (enabled: string[]) => enabled.map(e => `#extension ${e}â€‹ : enable`).join('\n');

/** Convert a bundle with a defined entry point to a definition for that attribute or type. */
export const bundleToAttribute = makeBundleToAttribute(toTypeString, toTypeArgs);

/** Convert a bundle to a definition for all its attributes. */
export const bundleToAttributes = makeBundleToAttributes(toTypeString, toTypeArgs);

// Simple whitespace removal
const minifyCode = (code: string) => {
  code = code.replace(/ +/g, ' ');
  code = code.replace(/\n+/g, '\n');
  return code;
};

/** ES/CommonJS Transpiler */
export const transpileGLSL = makeTranspile('glsl', 'glsl', loadModule, compressAST, minifyCode);

/** Templated literal syntax:

```tsx
glsl`...`
``` */
export const glsl = (literals: TemplateStringsArray, ...tokens: string[]) => {
  const code = zip(literals, tokens).flat();
  return loadModuleWithCache(code.join(''));
};

/** Format `number` as GLSL `float` */
export const float = (x: number) => {
  const s = x.toString();
  return (!s.match(/\./)) ? s + '.0' : 0;  
};
/** Format `number` as GLSL `uint` */
export const uint = (x: number) => Math.round(x).toString();
/** Format `number` as GLSL `int` */
export const int = (x: number) => Math.round(x).toString();
