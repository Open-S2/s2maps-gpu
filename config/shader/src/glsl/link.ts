import { Tree } from '@lezer/common';

import { defineConstants, defineEnables, loadModule, loadModuleWithCache, DEFAULT_CACHE } from './shader';
import { rewriteUsingAST } from './ast';
import { makeLinker, makeLinkCode, makeLinkBundle, makeLinkModule } from '../util/link';

import { GLSL_VERSION } from './constants';

// Override GLSL version/prefix
let PREAMBLE = `#version ${GLSL_VERSION}`;
export const setPreamble = (s: string): string => PREAMBLE = s;
export const getPreamble = (): string => PREAMBLE;
const getPreambles = (): string[] => [getPreamble()];

// No preprocessor renames
const getRenames = () => new Map<string, string>();

export const linker     = makeLinker(getPreambles, getRenames, defineConstants, defineEnables, rewriteUsingAST);
export const linkBundle = makeLinkBundle(linker);
export const linkModule = makeLinkModule(linker);
export const linkCode   = makeLinkCode(linker, loadModuleWithCache, DEFAULT_CACHE);
