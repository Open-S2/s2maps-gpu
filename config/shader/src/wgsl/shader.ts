import { Tree } from '@lezer/common';
import { ParsedModule, ParsedModuleCache, ShaderDefine } from './types';

import { makeLoadModule, makeLoadModuleWithCache } from '../util/shader';
import { makeBundleToAttribute, makeBundleToAttributes } from '../util/bundle';
import { makeTranspile } from '../util/transpile';

import { makeASTParser, compressAST, decompressAST } from './ast';
import { toTypeString, toTypeArgs } from './type';
import { removeComments, removeWhiteSpace, renameLocals } from './minify';
import { parser } from './grammar/wgsl';

import LRU from 'lru-cache';
import zip from 'lodash/zip';

export { loadStaticModule, loadVirtualModule, bindEntryPoint } from '../util/shader';

/** LRU cache for parsed shader code */
export const makeModuleCache = (options: Record<string, any> = {}) => new LRU<number, ParsedModule>({
  max: 100,
  ...options,
});

export const DEFAULT_CACHE = makeModuleCache();

/** Parse WGSL code into lezer tree */
export const parseShader = (code: string): Tree => parser.parse(code);

/** Parse a code module into its in-memory representation (AST + symbol table) */
export const loadModule = makeLoadModule(parseShader, makeASTParser, compressAST, decompressAST);

/** Use cache to load modules */
export const loadModuleWithCache = makeLoadModuleWithCache(loadModule, DEFAULT_CACHE);

/** Make WGSL constant definitions */
export const defineConstants = (defs: Record<string, ShaderDefine>): string => {
  const out = [];
  for (let k in defs) if (k[0] !== '@' && defs[k] != null) out.push(`const ${k} = ${defs[k]};`);
  return out.join("\n");
}

/** Make WGSL constant definitions */
export const defineEnables = (enabled: string[]) => enabled.length ? `enable ${enabled.join(', ')};` : '';

/** Convert a bundle with a defined entry point to a definition for that attribute or type. */
export const bundleToAttribute = makeBundleToAttribute(toTypeString, toTypeArgs);

/** Convert a bundle to a definition for all its attributes. */
export const bundleToAttributes = makeBundleToAttributes(toTypeString, toTypeArgs);

// Simple whitespace / comment removal
const minifyCode = (code: string) => {
  code = removeComments(code);
  code = renameLocals(code);
  code = removeWhiteSpace(code);
  return code;
};

/** ES/CommonJS Transpiler */
export const transpileWGSL = makeTranspile('wgsl', 'wgsl', loadModule, compressAST, minifyCode);

/** Templated literal syntax:

```tsx
wgsl`...`
``` */
export const wgsl = (literals: TemplateStringsArray, ...tokens: string[]) => {
  const code = zip(literals, tokens).flat();
  return loadModuleWithCache(code.join(''));
};

/** Format `number` as WGSL `f32` */
export const f32 = (x: number) => {
  const s = x.toString();
  return (!s.match(/\./)) ? s + '.0' : s;  
};
/** Format `number` as WGSL `u32` */
export const u32 = (x: number) => Math.round(x).toString() + 'u';
/** Format `number` as WGSL `i32` */
export const i32 = (x: number) => Math.round(x).toString();
