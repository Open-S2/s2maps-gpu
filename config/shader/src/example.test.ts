import { linkCode, linkModule, linkBundle } from './glsl/link';
import { loadModule } from './glsl/shader';

const trim = (s: string) => s.replace(/(^\s+|\s+$)/g, '');

describe("example linking", () => {

  it("links two GLSL shaders from source code using an import", () => {
    
    const color = trim(`
#pragma global
uniform vec4 color;

#pragma export
vec4 getColor() {
  return color;
}`);
    
    const main = trim(`
#pragma import { getColor } from 'color';

void main() {
  gl_FragColor = getColor();
}
`);
    
    const linked = linkCode(main, {color});
    expect(linked).toMatchSnapshot();
    
  });

  it("links two GLSL shaders from bundle using an import", () => {
    
    const color = trim(`
#pragma global
uniform vec4 color;

#pragma export
vec4 getColor() {
  return color;
}`);
    
    const main = trim(`
#pragma import { getColor } from 'path/to/getColor';

void main() {
  gl_FragColor = getColor();
}
`);

    // import main from 'path/to/main.glsl'
    // generated by @use-gpu/glsl-loader
    const bundle = {
      module: loadModule(main, 'main'),
      libs: {
        'path/to/getColor': {
          module: loadModule(color, 'getColor'),
          libs: {},
        }
      }
    };

    const linked = linkBundle(bundle);
    expect(linked).toMatchSnapshot();
    
  });
  
  it("links two GLSL shaders from source code using a prototype", () => {
    
    const getColor = trim(`
#pragma global
uniform vec4 color;

#pragma export
vec4 getColor() {
  return color;
}`);
    
    const main = trim(`
vec4 getColor();

void main() {
  gl_FragColor = getColor();
}`);

    const linked = linkCode(main, {}, {getColor});
    expect(linked).toMatchSnapshot();
  });
  
  
  it("links two GLSL shaders as modules using a prototype", () => {
    
    const getColor = loadModule(trim(`
#pragma global
uniform vec4 color;

#pragma export
vec4 getColor() {
  return color;
}`), 'getColor');
    
    const main = loadModule(trim(`
vec4 getColor();

void main() {
  gl_FragColor = getColor();
}`), 'main');

    const linked = linkModule(main, {}, {getColor});
    expect(linked).toMatchSnapshot();
  });

  it("readme example", () => {
    
    const color = trim(`
#pragma export
vec4 getColor() {
  return vec4(used(), 0.5, 0.0, 1.0);
}

float used() {
  return 1.0;
}

void unused() {
  // ...
}`);
    
    const main = trim(`
#pragma import { getColor } from 'path/to/color'

void main() {
  gl_FragColor = getColor();
}
`);

    // import main from 'path/to/main.glsl'
    // generated by @use-gpu/glsl-loader
    const bundle = {
      module: loadModule(main, 'main'),
      libs: {
        'path/to/color': {
          module: loadModule(color, 'color'),
        }
      }
    };

    const linked = linkBundle(bundle);
    expect(linked).toMatchSnapshot();
  });
});