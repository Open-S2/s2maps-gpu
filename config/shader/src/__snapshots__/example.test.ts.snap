// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`example linking links two GLSL shaders as modules using a prototype 1`] = `
"#version 450
uniform vec4 color;

vec4 _01_getColor() {
  return color;
}
vec4 _01_getColor();

void main() {
  gl_FragColor = _01_getColor();
}"
`;

exports[`example linking links two GLSL shaders from bundle using an import 1`] = `
"#version 450
uniform vec4 color;

vec4 _01_getColor() {
  return color;
}
void main() {
  gl_FragColor = _01_getColor();
}"
`;

exports[`example linking links two GLSL shaders from source code using a prototype 1`] = `
"#version 450
uniform vec4 color;

vec4 _01_getColor() {
  return color;
}
vec4 _01_getColor();

void main() {
  gl_FragColor = _01_getColor();
}"
`;

exports[`example linking links two GLSL shaders from source code using an import 1`] = `
"#version 450
uniform vec4 color;

vec4 _01_getColor() {
  return color;
}
void main() {
  gl_FragColor = _01_getColor();
}"
`;

exports[`example linking readme example 1`] = `
"#version 450
vec4 _01_getColor() {
  return vec4(_01_used(), 0.5, 0.0, 1.0);
}

float _01_used() {
  return 1.0;
}


void main() {
  gl_FragColor = _01_getColor();
}"
`;
